pages/posts/2017/12-december/140-continuous-integration-at-subvisual.md

```diff
@@ -0,0 +1,116 @@
+
+# Code Reviews
+
+We've been doing code reviews and enforcing them as required for all of our projects for a long time now. Our rule is simple: every change that is merged must be approved by at least one person other than yourself. So you either need someone else to review it, or you can pair with someone during development.
+
+And with our tendency for noticing every single detail, it's rare for a single extra newline to go unnoticed. 
+
+![2017-10-29-181705_790x539_scrot.png](https://s3.eu-west-1.amazonaws.com/subvisual/blog/posts/140/github1.png)
+We're extremely paranoid about the quality of our code.
+
+Recent updates to Github helped this a lot, with Code Reviews now being a part of the Pull Requests interface.
+
+![2017-10-29-181817_783x430_scrot.png](https://s3.eu-west-1.amazonaws.com/subvisual/blog/posts/140/github2.png)
+
+# CI builds
+
+One of the reasons we've chosen Semaphore when we first got started was their great ability to customize the setup and build process. We can create our own bash script to customize the build, and even install additional packages if we need to (especially useful if we're using edge versions of a library, and need to compile them from source). Here's an example for a heavily-customized Elixir/React application:
+
+```bash
+# setup
+source $HOME/.kerl/installs/19/activate
+nvm use $NODE_VERSION
+mix local.hex --force
+mix local.rebar --force
+mix deps.get
+yarn install
+MIX_ENV=test mix ecto.setup
+
+# build
+mix credo
+mix test --exclude feature
+mix test --only feature # trust us. there's a reason for this
+yarn lint
+```
+
+It also allows us to customize what directories get cached between builds.
+
+These two features combined make it a great resource even if your framework or language isn't yet officially supported by them. You can input the commands to download and compile such dependencies, and then cache them so that it doesn't slow down your future builds.
+
+This also ties in very well with another of [Subvisual's best practices]: for every single project, we strive to have a set of scripts (`bin/setup`, `bin/server`, and so on), that will do all the magic required to get a newcomer up and running, without even needing to know what dependencies are required, and how to install them.
+
+Since we already have these scripts built, we can often use them as our build commands in Semaphore.
+
+That's not what's happening in the above example though. You can blame that on my own laziness...
+
+# Automatic deploys to staging
+
+Most of our projects have two environments online. A staging server, often password-protected, allows us to test things on a controlled environment before getting them live to a real audience.
+
+The production server is where the magic happens.
```

Comment:
I have multiple issues with this sentence.

First, the phrasing. You say `The production server ...` but so far you haven't introduced that a production server exists, only the staging server. I suggest following on the previous sentence with something like `And a production server, ...`.

Second, is too short. I'm sure you can add something to it, like `It is the live version of the application that will be used by real users.`.

Lastly, magic also happens in staging, it's just fake magic. This is where the **real** magic happens :stuck_out_tongue:.

Reply:
all fixed